### PR TITLE
Add support for enumerable views in creme.form.Select2 widget

### PR DESCRIPTION
**Pull request by joehybird**
Refs #441

### CODE REVIEW

File: creme/creme_core/core/enumerable.py [hunk 1]
```diff
@@ -94,8 +95,33 @@ def _queryset(self):
...
+            map(self.instance_as_dict, queryset)
+        )
+
+        return list(islice(choices, limit) if limit else choices)
+
+    def filter_by_values(self, user, values, limit=None):
+        queryset = self._queryset().filter(id__in=values)
+        return list(
+            map(self.instance_as_dict, queryset[:limit] if limit else queryset)
+        )
```
Comment (genglert) [inline creme/creme_core/core/enumerable.py:113]: It seems this method could return a QuerySet, and `choices()` does the `map()` & `limit` work

File: creme/creme_core/core/enumerable.py [hunk 2]
```diff
@@ -94,8 +95,33 @@ def _queryset(self):
...
+    def filter_by_term(self, user, term, limit=None):
+        queryset = self._queryset()
+        term = term.lower()
+
+        choices = filter(
+            lambda c: term in c['label'].lower(),
+            map(self.instance_as_dict, queryset)
+        )
+
+        return list(islice(choices, limit) if limit else choices)
```
Comment (genglert) [inline creme/creme_core/core/enumerable.py:107]: I understand why you filter after having retrieved all the instance: you don't know how the label (i.e. `__str__` in general ) is built, so you build all the label then you filter.
But it has 2 big issues:
 - `term in c['label'].lower()` does not work well with diacritics I presume. It could be done in Python, so it's not the main issue (but I can give idea of sadistic unit tests...).
 - the filtering is not made by the SQL server. The returned JSON will be short (so it avoid a part of the performance issue), but it's clearly not satisfactory.

So we must found which field(s) should be used in order to make a '__icontains' query.
 - 1) If the related model gets a "_search_fields" attribute, use them.
 - 2) If not, search the first viewable CharField in the model (404/409/400 if no field is found of course)

I presume that our models could remain the same (i.e. not need to add the attribute "_search_fields") : the rule 2 will work perfectly, our model registered in CremeConfig have just a CharField title/name in 90% of the cases.
We prefix "_search_fields" with an underscore because it's clearly an "API" which we will break for a more powerful system in the future.
Notice that for problematic modes, people can still register their own Enumerator.

**Review by genglert — CHANGES_REQUESTED** [review]

File: creme/creme_core/forms/listview.py [hunk 3]
```diff
@@ -40,9 +40,12 @@
 from creme.creme_core.forms.base import CremeForm
 from creme.creme_core.forms.widgets import DatePickerMixin
 from creme.creme_core.models import Relation
-# from creme.creme_core.utils.dates import dt_from_str
 from creme.creme_core.utils.date_range import CustomRange
 
+# import unicodedata
+# from django.conf import settings
+# from django.utils.formats import get_format
+# from creme.creme_core.utils.dates import dt_from_str
```
[the remark this answers is not in the document]
Comment (joehybird) [reply creme/creme_core/forms/listview.py:48]: Yes I'm switching between eclipse & vscode all the time... I guess vscode did this

File: creme/creme_core/tests/views/test_listview.py [hunk 4]
```diff
@@ -1814,7 +1815,7 @@ def search(searched):
             )
             return self._get_lv_content(self._get_lv_node(response))
 
-        content = search(cat1.name[:5])  # Invalid we need an ID => no filter
+        content = search(999999)  # Invalid we need an ID => no filter
```
[the remark this answers is not in the document]
Comment (joehybird) [reply creme/creme_core/tests/views/test_listview.py:1818]: I didn't know this guy :+1:

File: creme/creme_core/tests/core/test_enumerable.py [hunk 5]
```diff
@@ -50,6 +50,32 @@ def test_basic_choices_fk(self):
...
+            for id, title in FakeCivility.objects.values_list('id', 'title')
+        ]
+
+        self.assertListEqual(expected[:2], enum.choices(user, limit=2))
+        self.assertListEqual(expected, enum.choices(user, limit=100))
+
+    def test_basic_choices_fk__term(self):
+        user = self.login()
+        registry = _EnumerableRegistry()
+        self.assertEqual('_EnumerableRegistry:', str(registry))
```
[the remark this answers is not in the document]
Comment (joehybird) [reply creme/creme_core/tests/core/test_enumerable.py:70]: Not exactly, here we filter the items. I should use 'Mi' and 'Miss' & 'Mister' as result.

File: creme/creme_core/tests/core/test_enumerable.py [hunk 6]
```diff
@@ -50,6 +50,32 @@ def test_basic_choices_fk(self):
         enum2 = registry.enumerator_by_field(field=field)
         self.assertEqual(expected, enum2.choices(user))
 
+    def test_basic_choices_fk__limit(self):
+        user = self.login()
```
[the remark this answers is not in the document]
Comment (joehybird) [reply creme/creme_core/tests/core/test_enumerable.py:54]: It could be merge with the previous test if you want

File: creme/creme_core/forms/listview.py [hunk 7]
```diff
@@ -193,6 +196,113 @@ def value_from_datadict(self, data, files, name):
...
+            )
+
+        return [*groups.items()]
+
+    def get_context(self, name, value, attrs):
+        context = super().get_context(name, value, attrs)
+        context['enum'] = {
+            'groups': self.group_choices(
+                self.enumerator.choices(values=[value] if value else None),
+                value
```
[the remark this answers is not in the document]
Comment (genglert) [reply creme/creme_core/forms/listview.py:296]: Sorry I just meant to use keyword arguments because the code is tricky & it's easier to understand

File: creme/creme_core/tests/core/test_enumerable.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (genglert) [reply creme/creme_core/tests/core/test_enumerable.py:54]: I reword it (sorry I was tired): is the login() useful ? It seems you just use the registry, not a view.

File: creme/creme_core/tests/core/test_enumerable.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (genglert) [reply creme/creme_core/tests/core/test_enumerable.py:70]: "self.assertEqual('_EnumerableRegistry:', str(registry))" => i do not understand this line ; & you made the same assertions several times.

File: creme/creme_core/tests/core/test_enumerable.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (joehybird) [reply creme/creme_core/tests/core/test_enumerable.py:54]: It is for the 'user' argument for choices. Not useful here but needed by the api to implement specific enumerators (I guess so)

File: creme/creme_core/tests/core/test_enumerable.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joehybird) [reply creme/creme_core/tests/core/test_enumerable.py:70]: copy/paste issue, I didn't pay attention to it :disappointed:

File: creme/creme_core/tests/core/test_enumerable.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (genglert) [reply creme/creme_core/tests/core/test_enumerable.py:54]: Ok ; you can use `self.create_user()` which is lighter than `self.login()` (it creates 2 users & one role).

File: creme/creme_core/enumerators.py [hunk 8]
```diff
@@ -24,6 +24,8 @@
 
 
 class UserEnumerator(enumerable.QSEnumerator):
+    search_fields = ('first_name', 'last_name', 'username')
```
Comment (genglert) [inline creme/creme_core/enumerators.py:27]: "_search_fields" ? (underscore)

File: .circleci/config.yml [hunk 9]
```diff
@@ -220,6 +220,10 @@ jobs:
       - create-creme-project
       - setup-creme-unit-tests:
           local_settings: 'pgsql_settings'
+      - run:
+          name: "PostgreSQL extensions"
+          command: |
+            echo "SELECT * FROM pg_available_extensions"|creme dbshell --settings=<< pipeline.parameters.instance_directory >>.settings
```
Comment (genglert) [inline .circleci/config.yml:226]: Would deserve its own commit I presume.

**Review by genglert — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 9 above)
Comment (joehybird) [reply .circleci/config.yml]: I moved it to [draft/pg_enumerable_unaccent](https://github.com/HybirdCorp/creme_crm/tree/draft/pg_enumerable_unaccent)

**Review by genglert — APPROVED** [review]
